lib/lhm/sql_retry.rb

```diff
@@ -15,26 +16,128 @@ module Lhm
   # https://github.com/kamui/retriable. Additionally, a "log_prefix" option,
   # which is unique to SqlRetry can be used to prefix log output.
   class SqlRetry
-    def initialize(connection, options = {})
+    RECONNECT_SUCCESSFUL_MESSAGE = "LHM successfully reconnected to initial host:"
+    CLOUDSQL_VERSION_COMMENT = "(Google)"
+
+    MYSQL_VAR_NAMES = {
+      hostname: "@@global.hostname",
+      server_id: "@@global.server_id",
+      version_comment: "@@version_comment",
+    }
+
+    # This internal error is used to trigger retries from the parent Retriable.retriable in #with_retries
+    class ReconnectToHostSuccessful < Lhm::Error; end
+
+    def initialize(connection, options = {}, with_consistent_host = true)
       @connection = connection
+      @log_prefix = options.delete(:log_prefix)
       @global_retry_config = default_retry_config.dup.merge!(options)
+      if (@with_consistent_host = with_consistent_host)
+        @initial_hostname = hostname
+        @initial_server_id = server_id
+      end
     end
 
+    # Complete explanation of algorithm: https://github.com/Shopify/db-engineering/issues/98#issuecomment-934948590
     def with_retries(retry_config = {})
-      cnf = @global_retry_config.dup.merge!(retry_config)
-      @log_prefix = cnf.delete(:log_prefix) || "SQL Retry"
-      Retriable.retriable(cnf) do
+      # Overrides log prefix if necessary or passed from parent to child instance (ex: Chunker -> ChunkInsert)
+      old_prefix = @log_prefix
+      @log_prefix = retry_config.delete(:log_prefix)
+
+      retry_config = @global_retry_config.dup.merge!(retry_config)
+
+      Retriable.retriable(retry_config) do
+        if @with_consistent_host
+          raise Lhm::Error.new("Could not reconnected to initial MySQL host. Aborting to avoid data-loss") unless same_host_as_initial?
+        end
+
         yield(@connection)
+      rescue StandardError => e
+        # Not all errors should trigger a reconnect. Some errors such be raised and abort the LHM (such as reconnecting to the wrong host).
+        raise e unless error_can_trigger_reconnect?(e)
+        reconnect_with_host_check! if @with_consistent_host
       end
+    ensure
+      # Restore the initial log prefix once outside of the block (ex: Chunker -> ChunkInsert).
+      @log_prefix = old_prefix
     end
 
     attr_reader :global_retry_config
 
     private
 
+    def hostname
+      mysql_single_value(MYSQL_VAR_NAMES[:hostname])
+    end
+
+    def server_id
+      mysql_single_value(MYSQL_VAR_NAMES[:server_id])
+    end
+
+    def cloudsql?
+      mysql_single_value(MYSQL_VAR_NAMES[:version_comment]).include?(CLOUDSQL_VERSION_COMMENT)
+    end
+
+    def mysql_single_value(name)
+      query = Lhm::ProxySQLHelper.tagged("SELECT #{name} LIMIT 1")
```

Comment:
@connection is an ActiveRecord connection to not create a loop, since `Lhm::Connection` calls `with_retries`. Therefore, they need to be tagged manually.